memclear.c

```diff
@@ -90,30 +110,125 @@ EFI_STATUS efi_main(EFI_HANDLE image_handle, EFI_SYSTEM_TABLE* system_table)
 				continue;
 			}
 
-			debug_hex(L"Start: 0x", start);
-			debug_hex(L"End:   0x", end);
-
+			//debug_hex(L"Cleaning the conventional memory region: 0x", start);
 			/* Clear one page at a time */
 			for (i = 0; i < Map->NumberOfPages; i++, addr += PAGE_SIZE) {
 				gBS->SetMem(addr, PAGE_SIZE, 0);
 			}
 
 			gBS->FreePages(start, Map->NumberOfPages);
+			//debug_hex(L"Cleaned the conventional memory region: 0x", start);
+		}
+		else {
+
+			int Clearable = 0;
+			switch (Map->Type) {
+				// Becareful: The following mapped regions will be cleared
+				case EfiACPIReclaimMemory:
+				case EfiACPIMemoryNVS:
+				case EfiMemoryMappedIO:
+				case EfiMemoryMappedIOPortSpace:
+					Clearable = 1;
+					break;
+				case EfiBootServicesData: 
+					if (rsp < start || rsp > end) {
+						Clearable = 1;
+						break;
+					}
+				case EfiBootServicesCode:
+					if (rip < start || rip > end) {
+						Clearable = 1;
+						break;
+					}
```

Comment:
This will clear UEFI code and data sections while you're still using them. I'm surprised it works for you :). I would expect that once you return from memclear.efi (or you get an interrupt), everything will explode.